### PR TITLE
MINOR: use 'mapKey' to avoid unnecessary grouped data

### PR DESCRIPTION
**Pull request by chia7712**
as title. Thanks to great KRPC :)

# changes in this PR

1. add 'mapKey=true' to `DescribeLogDirsRequest`
1. rename `PartitionIndex` to `Partitions` for `DescribeLogDirsRequest`
1. add 'mapKey=true' to `ElectLeadersRequest`
1. rename `PartitionId` to `Partitions` for `ElectLeadersRequest`
1. add 'mapKey=true' to `ConsumerProtocolAssignment`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/resources/common/message/ElectLeadersRequest.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     { "name": "TopicPartitions", "type": "[]TopicPartitions", "versions": "0+", "nullableVersions": "0+",
       "about": "The topic partitions to elect leaders.",
       "fields": [
-        { "name": "Topic", "type": "string", "versions": "0+", "entityType": "topicName",
+        { "name": "Topic", "type": "string", "versions": "0+", "entityType": "topicName", "mapKey": true,
           "about": "The name of a topic." },
         { "name": "PartitionId", "type": "[]int32", "versions": "0+",
```
Comment (dajac) [inline clients/src/main/resources/common/message/ElectLeadersRequest.json]: nit: While we are here, should we rename this field to `Partitions`? I was confused by this while reading the code. When I read `PartitionId`, I don't really expect an array.

Comment (chia7712) [reply clients/src/main/resources/common/message/ElectLeadersRequest.json]: make sense

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 2]
```diff
@@ -2581,10 +2581,10 @@ public DescribeReplicaLogDirsResult describeReplicaLogDirs(Collection<TopicParti
                 List<Integer> partitionIndex = new ArrayList<>();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: Change this variable name too?

**Comment by ijuma** [comment]
Looks like ConsumerProtocolAssignment was also changed to have mapKey, so worth updating the PR to include that information.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 2 above)
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: I do agree.

**Review by dajac — APPROVED** [review]
@chia7712 LGTM besides @ijuma's comment. It would be good to rename that variable to be consistent.

**Comment by chia7712** [comment]
> Looks like ConsumerProtocolAssignment was also changed to have mapKey, so worth updating the PR to include that information.

done

**Comment by chia7712** [comment]
the error is traced by #10024

**Pull request merged**